Lcm call matlab

RussTedrake opened the pull request:
Provides an interface for calling matlab over lcm (e.g. for plotting).
Partial solution to #3647.

Hopefully the documentation in `lcm_call_matlab.h` describes it all.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4256)
<!-- Reviewable:end -->

hongkai-dai commented:
My MATLAB crashed when I run "lcm_cal_matlab_client". I am on a Ubuntu 14.04 machine with MATLAB R2015a. The message is
```
Configuration:
  Crash Decoding      : Disabled
  Crash Mode          : continue (default)
  Current Graphics Driver: Unknown hardware
  Current Visual      : 0x21 (class 4, depth 24)
  Default Encoding    : UTF-8
  GNU C Library       : 2.19 stable
  Host Name           : jiahui-All-Series
  MATLAB Architecture : glnxa64
  MATLAB Root         : /usr/local/MATLAB/R2015a
  MATLAB Version      : 8.5.0.197613 (R2015a)
  OpenGL              : hardware
  Operating System    : Linux 3.13.0-101-generic #148-Ubuntu SMP Thu Oct 20 22:08:32 UTC 2016 x86_64
  Processor ID        : x86 Family 6 Model 60 Stepping 3, GenuineIntel
  Virtual Machine     : Java 1.7.0_60-b19 with Oracle Corporation Java HotSpot(TM) 64-Bit Server VM mixed mode
  Window System       : The X.Org Foundation (11501000), display :0

Fault Count: 1


Abnormal termination:
abort()

Register State (from fault):
  RAX = 0000000000000000  RBX = 0000000000000059
  RCX = ffffffffffffffff  RDX = 0000000000000006
  RSP = 00007ffb57727f28  RBP = 00007ffb57728240
  RSI = 000000000000630b  RDI = 00000000000062d5

   R8 = 0000000000000000   R9 = 00007ffb6c89d817
  R10 = 0000000000000008  R11 = 0000000000000202
  R12 = 00007ffb577280b0  R13 = 0000000000000005
  R14 = 0000000000000059  R15 = 0000000000000005

  RIP = 00007ffb6c830c37  EFL = 0000000000000202

   CS = 0033   FS = 0000   GS = 0000

Stack Trace (from fault):
[  0] 0x00007ffb6c830c37                    /lib/x86_64-linux-gnu/libc.so.6+00224311 gsignal+00000055
[  1] 0x00007ffb6c834028                    /lib/x86_64-linux-gnu/libc.so.6+00237608 abort+00000328
[  2] 0x00007ffb6c86d2a4                    /lib/x86_64-linux-gnu/libc.so.6+00471716
[  3] 0x00007ffb6c904bbc                    /lib/x86_64-linux-gnu/libc.so.6+01092540 __fortify_fail+00000092
[  4] 0x00007ffb6c903a90                    /lib/x86_64-linux-gnu/libc.so.6+01088144
[  5] 0x00007ffb6c904b07                    /lib/x86_64-linux-gnu/libc.so.6+01092359
[  6] 0x00007ffac82b72e9 /home/hongkai/drake-distro/build/install/lib/liblcm.so.1+00021225 lcm_handle_timeout+00000057
[  7] 0x00007ffac86daae0 /home/hongkai/drake-distro/drake/matlab/util/lcm_call_matlab_client.mexa64+00006880 mexFunction+00000416
[  8] 0x00007ffb6241dcea     /usr/local/MATLAB/R2015a/bin/glnxa64/libmex.so+00150762 mexRunMexFile+00000090
[  9] 0x00007ffb6241aa74     /usr/local/MATLAB/R2015a/bin/glnxa64/libmex.so+00137844
[ 10] 0x00007ffb6241b684     /usr/local/MATLAB/R2015a/bin/glnxa64/libmex.so+00140932
[ 11] 0x00007ffb61753aea /usr/local/MATLAB/R2015a/bin/glnxa64/libmwm_dispatcher.so+00813802 _ZN8Mfh_file16dispatch_fh_implEMS_FviPP11mxArray_tagiS2_EiS2_iS2_+00000762
[ 12] 0x00007ffb61753fb0 /usr/local/MATLAB/R2015a/bin/glnxa64/libmwm_dispatcher.so+00815024 _ZN8Mfh_file11dispatch_fhEiPP11mxArray_tagiS2_+00000032
[ 13] 0x00007ffb609c3d20 /usr/local/MATLAB/R2015a/bin/glnxa64/libmwm_interpreter.so+04189472
[ 14] 0x00007ffb609444b3 /usr/local/MATLAB/R2015a/bin/glnxa64/libmwm_interpreter.so+03667123
[ 15] 0x00007ffb609746ee /usr/local/MATLAB/R2015a/bin/glnxa64/libmwm_interpreter.so+03864302
[ 16] 0x00007ffb6097b597 /usr/local/MATLAB/R2015a/bin/glnxa64/libmwm_interpreter.so+03892631
[ 17] 0x00007ffb60976cff /usr/local/MATLAB/R2015a/bin/glnxa64/libmwm_interpreter.so+03874047
[ 18] 0x00007ffb60977934 /usr/local/MATLAB/R2015a/bin/glnxa64/libmwm_interpreter.so+03877172
[ 19] 0x00007ffb609ed2ce /usr/local/MATLAB/R2015a/bin/glnxa64/libmwm_interpreter.so+04358862
[ 20] 0x00007ffb61753c39 /usr/local/MATLAB/R2015a/bin/glnxa64/libmwm_dispatcher.so+00814137 _ZN8Mfh_file16dispatch_fh_implEMS_FviPP11mxArray_tagiS2_EiS2_iS2_+00001097
```
My command is just
```
$ cd DRAKE-DISTRO/drake
$ matlab -nosplash -nodesktop
```
And in MATLAB terminal, I did
```
addpath_drake
lcm_call_matlab_client
```

---

Reviewed 5 of 10 files at r1, 3 of 5 files at r2.
Review status: 8 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/matlab/util/lcm_call_matlab_client.cc, line 19 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXWyQ99wxoihAgpHbmR:-KXWyQ99wxoihAgpHbmS:bmsppk3) ([raw file](https://github.com/robotlocomotion/drake/blob/6574cc720f60db8be302d619fdd22a4de4bf7d53/drake/matlab/util/lcm_call_matlab_client.cc#L19)):*
> ```cpp
> 
> void flushMatlabEventBuffer(void) {
>   // An ugly hack.  drawnow used to work, but the internet confirms that it
> ```

Mathworks documentation thinks drawnow works in both R2015 and R2016
https://www.mathworks.com/help/releases/R2015a/matlab/ref/drawnow.html
https://www.mathworks.com/help/releases/R2015b/matlab/ref/drawnow.html
https://www.mathworks.com/help/releases/R2016a/matlab/ref/drawnow.html
https://www.mathworks.com/help/matlab/ref/drawnow.html

---

*[drake/matlab/util/lcm_call_matlab_client.cc, line 41 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXWzjGgPRCxaPABtt1D:-KXWzjGgPRCxaPABtt1E:bubt7nx) ([raw file](https://github.com/robotlocomotion/drake/blob/6574cc720f60db8be302d619fdd22a4de4bf7d53/drake/matlab/util/lcm_call_matlab_client.cc#L41)):*
> ```cpp
>     for (i = 0; i < msg->nrhs; i++) {
>       if (msg->rhs[i].num_bytes == 0) {
>         mexPrintf("rhs %d seems to have zero bytes.  dropping message %s.\n", i,
> ```

This zero bytes rhs is not tested.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-:-KXX3YLtnG7Qlrdal6tf:b-a72cym)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
Review status: 8 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/matlab/util/lcm_call_matlab_client.cc, line 19 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXWyQ99wxoihAgpHbmR:-KXYgHKmzhWzWrx4KwJt:blyaywx) ([raw file](https://github.com/robotlocomotion/drake/blob/6574cc720f60db8be302d619fdd22a4de4bf7d53/drake/matlab/util/lcm_call_matlab_client.cc#L19)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> Mathworks documentation thinks drawnow works in both R2015 and R2016
> https://www.mathworks.com/help/releases/R2015a/matlab/ref/drawnow.html
> https://www.mathworks.com/help/releases/R2015b/matlab/ref/drawnow.html
> https://www.mathworks.com/help/releases/R2016a/matlab/ref/drawnow.html
> https://www.mathworks.com/help/matlab/ref/drawnow.html
</details>
FYI , `drawnow` should definitely clear the event queue. If you have a repro case where it doesn't, let me know.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 8 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/matlab/util/lcm_call_matlab_client.cc, line 19 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXWyQ99wxoihAgpHbmR:-KXYkoI8zTRwY00eupme:b1hy5a7) ([raw file](https://github.com/robotlocomotion/drake/blob/6574cc720f60db8be302d619fdd22a4de4bf7d53/drake/matlab/util/lcm_call_matlab_client.cc#L19)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary>
> FYI , `drawnow` should definitely clear the event queue. If you have a repro case where it doesn't, let me know.
</details>
it absolutely did not in all of my testing (on mac in R2015b).  the pause did the trick 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Debugged with Hongkai on slack last night, and realized that this is my old nemesis:
https://github.com/openhumanoids/oh-distro-private/issues/644
https://github.com/RobotLocomotion/drake/issues/1332

Long story short: both libpcre and libpoco both define the symbol g_regex_new().  When lcm c/c++ is run inside of matlab (>2014a), the matlab library order makes lcm get poco instead of pcre... which results in the crash.

RussTedrake commented:
For those who might not have access to open-humanoids, here is the summary from @mattantone:

This looks like a problem with library loading order.

Here's a minimal mex sample that breaks lcm in R2014b:
```
#include <mex.h>
#include <iostream>
#include <lcm/lcm-cpp.hpp>
#include <lcmtypes/drc/robot_state_t.hpp>

struct Listener {
    void onMessage(const lcm::ReceiveBuffer* rbuf,
              const std::string& chan,
              const drc::robot_state_t* msg) {
    }
};

void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[]) {
    lcm::LCM theLcm;
    Listener listener;
    theLcm.subscribe("ROBOT_STATE", &Listener::onMessage, &listener);
    theLcm.handle();
    std::cout << "HANDLED MESSAGE" << std::endl;
}
```

When lcm's subscribe method is invoked, it calls g_regex_new() to parse any regular expressions in the channel name. glib in turn calls pcre_compile2 (in libpcre), which then processes the regular expression string and normally returns with no error code (this is the case in R2014a, for instance, and in all non-matlab usages). But in R2014b, an error is returned by the g_regex_new() call, and lcm cannot proceed.

I compiled lcm in debug mode, attached gdb to a running matlab instance, and ran the above mex function, setting a breakpoint at lcm.c:321, inside of the subscribe method. Doing instruction-level stepping through execution revealed that in R2014a, pcre_compile2 is called correctly in libpcre as expected, while in R2014b, glib actually calls a version of pcre_compile2 that resides in a completely different shared object (!!) - namely libPocoFoundation, which apparently has a method with identical signature. Matlab includes this library in the bin/ directory of R2014b and presumably it's loaded at startup.

So instead of mex -> lcm -> glib -> pcre, we have mex -> lcm -> glib -> PocoFoundation, and apparently the two versions of pcre_compile2 set error codes differently. This seems to imply that lcm c libraries are incompatible with matlab R2014b (and probably newer).

Not sure how to fix this - perhaps there's a way to force library loading order. Or lcm could be compiled statically. Other ideas?

RussTedrake commented:
proposed solution:  I add a caveat (or even a runtime check) saying that it does not support matlab > 2014a on linux, and we merge.  Then, with the help of @RobotLocomotion/kitware , we resolve the issue on ubuntu in a future PR... most likely by linking this mex file against lcm-static instead of lcm.  (I would do it myself, but it looks like the lcm is not currently exporting pkgconfig or cmake files for the static targets.)  @hongkai-dai -- what do you think?

hongkai-dai commented:
I like that! Can't wait to call MATLAB in C++ using this PR!

I am still not very sure about the `drawnow` issue, but otherwise <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: 8 of 11 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-:-KXauhF9_-NsSfTxjP26:b2oathf)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
BTW, I think an old version of MATLAB < 2014 had issue to build sedumi, so the user need to turn `WITH_SEDUMI`off.

---

Review status: 8 of 11 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-:-KXax05DfukvBCdNYblu:b-fbz5fr)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
That's the type of reason why I thought it's better to do a runtime check than a guard in cmake.  I suspect one use case is that people build with a newer version of matlab in their path, but then just launch and old version of matlab to run this one app.  This will also be better for when we redistribute the nightly binaries.

---

Review status: 8 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/matlab/util/lcm_call_matlab_client.cc, line 41 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXWzjGgPRCxaPABtt1D:-KXb1H2xAxTfXe2TVmU8:b-5enayv) ([raw file](https://github.com/robotlocomotion/drake/blob/6574cc720f60db8be302d619fdd22a4de4bf7d53/drake/matlab/util/lcm_call_matlab_client.cc#L41)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> This zero bytes rhs is not tested.
</details>
it should never happen in proper use and I don't consider it part of the API.  it was useful for debugging, and i think it's better to leave it there than to zap it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-:-KXb0oigiN_b9GyJjEtp:b-uuytso)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
pushed.

---

Review status: 8 of 11 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-:-KXb1YkhVdkRsywcEren:bvwhisw)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 1 of 10 files at r1.
Review status: 9 of 11 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 1 of 1 files at r3.
Review status: 10 of 11 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
What is our minimum supported version of MATLAB? Getting LCM and SeDuMi working would not be especially difficult or time consuming, but so far we have not been explicitly testing or working around version-specific issues with MATLAB.

---

Review status: 10 of 11 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-:-KXfQTeeAADT9uhzPJ4R:b-ilz3zv)*
<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
> Or lcm could be compiled statically.

It's not clear to me why this is expected to solve the problem? It's still going to call out to glib (probably not static) which is still going to call out to the non-local symbol `pcre_compile2`... unless you propose to also have static glib using static libpcre?

Anyway, why is PocoFoundation exporting libpcre symbols? This seems like the real bug... (...and this is *why we have hidden symbol visibility!*)

> it looks like the lcm is not currently exporting pkgconfig or cmake files for the static targets

Arf, you're right :cry:... should be an easy fix, at least for CMake.

mwoehlke-kitware commented:
>> it looks like the lcm is not currently exporting pkgconfig or cmake files for the static targets
>
> Arf, you're right :cry:... should be an easy fix, at least for CMake.

Actually, this is only true for using LCM from its build directory; *installed* LCM does export `lcm-static`. Anyway, fixed by lcm-proj/lcm#148.

RussTedrake commented:
@mwoehlke-kitware -- you're right, it would have to be static all of the way down.  if lcm-static is not (i admit i wasn't sure), then it won't solve the problem.  what's your recommendation?

jwnimmer-tri commented:
Ideas:
- (1) In lcm_call_matlab_client, use fully-static versions of LCM, GLib, PCRE, etc all the way down to (but not including) libc.
- (2) Patch LCM upstream to offer a knob to disable regexp support and remove the PCRE dependency.
- (3) Rewrite lcm_call_matlab_client in Java.

mwoehlke-kitware commented:
- (4) Fix PocoFoundation?

Otherwise, I think I'd be partial to (3)... LCM does have Java bindings. (1) seems like it would be awkward at best, and I'm not sure if (2) would make sense for LCM upstream.

RussTedrake commented:
I had thought about (3), the java option, but i don't know that it's actually viable.  Matlab can load and call java, but I don't think you can actually call matlab from java (which is what I need here).  Plus the matlab-java bindings have pretty poor performance; we've switched a number of java lcm coders to c++ in the past because of this.

jamiesnape commented:
You can call MATLAB from Java, but I doubt the performance is good:
https://www.mathworks.com/help/matlab/matlab-engine-api-for-java.html

RussTedrake commented:
+@sherm1 for platform review, please.  we need to resolve the matlab/lcm issue, but the proposal is to (squash) merge this as is (the runtime guards are in place).  

---

Review status: 10 of 11 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-:-KXgeThhQ5jGFTOQULCI:bq1onh0)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed. A few comments below, PTAL. With minor fixes: <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.

---

Reviewed 6 of 10 files at r1, 4 of 5 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/CMakeLists.txt, line 89 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXh5K2oue5ellpmlLaf:-KXh5K2oue5ellpmlLag:b-dpxz1p) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/CMakeLists.txt#L89)):*
> ```CMake
>   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Werror=shadow")
>   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Werror=inconsistent-missing-override")
>   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wsign-compare")
> ```

BTW, why turn this on only for clang? g++ has the same flag.

---

*[drake/lcm/lcm_call_matlab.cc, line 19 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXhEQsfFBjdCuY2tp88:-KXhEQsfFBjdCuY2tp89:b-abmnrj) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcm/lcm_call_matlab.cc#L19)):*
> ```cpp
> LcmMatlabRemoteVariable::LcmMatlabRemoteVariable()
>     : uid(g_uid++)
> // TODO(russt): replace this with a random int64_t, e.g.
> ```

BTW, why does this need to be randomized?

---

*[drake/lcm/lcm_call_matlab.h, line 44 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXh6NSIsu1u3qaCmz12:-KXh6NSIsu1u3qaCmz13:b-iby7e3) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcm/lcm_call_matlab.h#L44)):*
> ```objc
> 
> /// Serialize our favorite data types into the lcm_matlab_array structure.
> /// To support a calling lcm_call_matlab a new type of input, simply implement
> ```

BTW, mangled sentence.

---

*[drake/lcm/lcm_call_matlab.h, line 84 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXh6wQcbagTSyk1WILL:-KXh6wQcbagTSyk1WILM:brbhm9z) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcm/lcm_call_matlab.h#L84)):*
> ```objc
> /// Invokes a mexCallMATLAB call on the remote client.
> ///
> /// @param channel The name of the channel to publish the LCM message on.
> ```

This parameter doesn't appear in the signature.

---

*[drake/lcm/lcm_call_matlab.h, line 88 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXh8cEnV7Lchi3xehSo:-KXh8cEnV7Lchi3xehSp:b-descya) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcm/lcm_call_matlab.h#L88)):*
> ```objc
> ///     that could have been passed to mexCallMATLAB is allowed.
> ///     https://www.mathworks.com/help/matlab/apiref/mexcallmatlab.html
> /// @param num_outputs Number of return variables (left-side arguments) to
> ```

num_outputs comes before function_name in the signature.

---

*[drake/lcm/lcm_call_matlab.h, line 91 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXh9fuiWL5CfWgfTshj:-KXh9fuiWL5CfWgfTshk:b-cbwb5e) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcm/lcm_call_matlab.h#L91)):*
> ```objc
> ///     request from matlab.  As in matlab, you need not request all of the
> ///     outputs for any given method.
> /// @option argument1 Any data type which has a ToLcmMatlabArray method
> ```

BTW, I don't recognize `@option` as a doxygen command -- is that a local macro?

---

*[drake/lcmtypes/lcmt_call_matlab.lcm, line 7 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXhBD_cpyzITfhjy6LK:-KXhBD_cpyzITfhjy6LL:b-8lh09f) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcmtypes/lcmt_call_matlab.lcm#L7)):*
> ```
>  *  (see http://www.mathworks.com/help/matlab/apiref/mexcallmatlab.html)
>  * but with output arguments assigned in the remote client's workspace
>  * due to the asynchronous nature of the protocol
> ```

BTW, obligatory complaint about comments needing to have sentence punctuation (here and below).

---

*[drake/lcmtypes/lcmt_call_matlab.lcm, line 13 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXhBcKWhJL5bIiPs-68:-KXhBcKWhJL5bIiPs-69:b-a8610x) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcmtypes/lcmt_call_matlab.lcm#L13)):*
> ```
> {
>   int32_t nlhs;                           // the number of output arguments
>   int64_t lhs[nlhs];                      // unique id for variables held in the MATLAB client's workspace
> ```

BTW, obligatory complaint about line length (here and below).

---

*[drake/matlab/util/lcm_call_matlab_client.cc, line 53 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXhCfiO1nukgpZPmVJ_:-KXhCfiO1nukgpZPmVJa:b9yiyit) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/matlab/util/lcm_call_matlab_client.cc#L53)):*
> ```cpp
> }
> 
> class Handler {
> ```

BTW, if this is intended as a local class it should be in an anonymous namespace. As written it is visible externally.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-:-KXhE_b_YVzKqviSHFDo:b-ar14h)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
I only skimmed the one h/cc pair, but two comments...

---

Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/lcm/lcm_call_matlab.cc, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXhKXx9Qsp7LCdXDTwq:-KXhKXxAL8VDv8H7z7o3:bsl5usi) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcm/lcm_call_matlab.cc#L66)):*
> ```cpp
> 
> void internal::PublishLcmCallMatlab(const drake::lcmt_call_matlab& msg) {
>   static ::lcm::LCM lcm;  // just keep a local copy here for publishing
> ```

Non-POD statics are not allowed per https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

---

*[drake/lcm/lcm_call_matlab.h, line 47 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXhKh1MiCp6HXdp6c7-:-KXhKh1MiCp6HXdp6c70:b-n1odur) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcm/lcm_call_matlab.h#L47)):*
> ```objc
> //  another one of these methods.
> 
> extern void ToLcmMatlabArray(const LcmMatlabRemoteVariable& var,
> ```

BTW META Are the `extern`s here doing anything?  I think they are superflous (for functions).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-:-KXhLKh0sLPOt5L5qr_L:b-gicwa7)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/lcm/test/lcm_call_matlab_test.cc, line 9 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXhQSZFRaq3HVG86smO:-KXhQSZGOi56VkLU00F8:b-4bjd7n) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcm/test/lcm_call_matlab_test.cc#L9)):*
> ```cpp
> // Note: Unfortunately these really only test whether the code compiles and
> // runs... the actual output must currently be verified by human inspection in
> // the matlab window.
> ```

BTW It would be straightforward to use the `DrakeMockLcm` to unit test that at least the correct LCM messages were being sent.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[drake/CMakeLists.txt, line 89 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXh5K2oue5ellpmlLaf:-KXhPJ3hvJIaUV6SprdE:bdjz9y2) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/CMakeLists.txt#L89)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, why turn this on only for clang? g++ has the same flag.
</blockquote></details>

i thought it was on for g++, because without it I was getting rejected by the gcc build servers (without the ability to reproduce locally with clang)?

---

*[drake/lcm/lcm_call_matlab.cc, line 19 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXhEQsfFBjdCuY2tp88:-KXhQ19ALCqoG00pNjAB:brgqazj) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcm/lcm_call_matlab.cc#L19)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, why does this need to be randomized?
</blockquote></details>

as it is, if we have multiple apps all publishing to the same viewer, their "remote local variables" will overwrite each other.

---

*[drake/lcm/lcm_call_matlab.cc, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXhKXx9Qsp7LCdXDTwq:-KXhRL6QQqAWI6-5u0_I:bj9hla7) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcm/lcm_call_matlab.cc#L66)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Non-POD statics are not allowed per https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables
</blockquote></details>

Done.  Per the recommendations in the style guide, replaced with a raw pointer (who's destructor will never be called)  and documented this decision.

---

*[drake/lcm/lcm_call_matlab.h, line 44 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXh6NSIsu1u3qaCmz12:-KXhT4YmWZT3BWUXaVIF:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcm/lcm_call_matlab.h#L44)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, mangled sentence.
</blockquote></details>

Done.

---

*[drake/lcm/lcm_call_matlab.h, line 47 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXhKh1MiCp6HXdp6c7-:-KXhT5KDeUMIMQcTrEyP:b-m4r5xy) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcm/lcm_call_matlab.h#L47)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW META Are the `extern`s here doing anything?  I think they are superflous (for functions).
</blockquote></details>

done.  good catch.  old habits die hard.

---

*[drake/lcm/lcm_call_matlab.h, line 84 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXh6wQcbagTSyk1WILL:-KXhTH0PhocGDZY2zrVT:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcm/lcm_call_matlab.h#L84)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

This parameter doesn't appear in the signature.
</blockquote></details>

Done.

---

*[drake/lcm/lcm_call_matlab.h, line 91 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXh9fuiWL5CfWgfTshj:-KXhTLPITLhG0Bde25SM:b524eas) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcm/lcm_call_matlab.h#L91)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, I don't recognize `@option` as a doxygen command -- is that a local macro?
</blockquote></details>

done.  hah!  good catch.  that was a flag we added to our matlab doxygen parsing.  

---

*[drake/lcm/test/lcm_call_matlab_test.cc, line 9 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXhQSZFRaq3HVG86smO:-KXhTx4Z_fu5u-jVVSri:b-7yd1q4) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcm/test/lcm_call_matlab_test.cc#L9)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW It would be straightforward to use the `DrakeMockLcm` to unit test that at least the correct LCM messages were being sent.
</blockquote></details>

Added a TODO.

---

*[drake/lcmtypes/lcmt_call_matlab.lcm, line 7 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXhBD_cpyzITfhjy6LK:-KXhUG1X_flIpo6dS8ZY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcmtypes/lcmt_call_matlab.lcm#L7)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, obligatory complaint about comments needing to have sentence punctuation (here and below).
</blockquote></details>

Done.

---

*[drake/lcmtypes/lcmt_call_matlab.lcm, line 13 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXhBcKWhJL5bIiPs-68:-KXhUM98KGt8NQQf7QnI:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcmtypes/lcmt_call_matlab.lcm#L13)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, obligatory complaint about line length (here and below).
</blockquote></details>

Done.

---

*[drake/matlab/util/lcm_call_matlab_client.cc, line 53 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXhCfiO1nukgpZPmVJ_:-KXhVi7F-FMTWFkSRZOE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/matlab/util/lcm_call_matlab_client.cc#L53)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW, if this is intended as a local class it should be in an anonymous namespace. As written it is visible externally.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 6 of 11 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[drake/CMakeLists.txt, line 89 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXh5K2oue5ellpmlLaf:-KXhWv8eYdpyvNxDPlSx:bua9z4u) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/CMakeLists.txt#L89)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

i thought it was on for g++, because without it I was getting rejected by the gcc build servers (without the ability to reproduce locally with clang)?
</blockquote></details>

@sherm1 On GCC, the `-Werror=all` above implies `-Wsign-compare`, which is why it doesn't have to be added separate for GCC.

---

*[drake/CMakeLists.txt, line 89 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXhX2yqYHc3FFov2BSH:-KXhX2yqYHc3FFov2BSI:bpls1iw) ([raw file](https://github.com/robotlocomotion/drake/blob/257b15358c8fcafbdff96c277cf73ecd2ccd8909/drake/CMakeLists.txt#L89)):*
> ```CMake
>   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Werror=shadow")
>   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Werror=inconsistent-missing-override")
>   set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wsign-compare")
> ```

To have any useful effect, this should be `-Werror=sign-compare`, I think?  A console-only warning doesn't help us (is ignored by CI)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 5 of 11 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/CMakeLists.txt, line 89 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXhX2yqYHc3FFov2BSH:-KXhYxjTGkOkQNh_DRKy:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/257b15358c8fcafbdff96c277cf73ecd2ccd8909/drake/CMakeLists.txt#L89)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

To have any useful effect, this should be `-Werror=sign-compare`, I think?  A console-only warning doesn't help us (is ignored by CI)?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r5.
Review status: 6 of 11 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 6 of 11 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[drake/lcm/lcm_call_matlab.h, line 88 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-KXh8cEnV7Lchi3xehSo:-KXh_ST2HCUXFTXtG8eX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8b55036ac0307a91fc00df919c8e38e4bf0f8929/drake/lcm/lcm_call_matlab.h#L88)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

num_outputs comes before function_name in the signature.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

---

Review status: 6 of 11 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4256#-:-KXhjrGS6132djEJGmCI:b-uwkowd)*
<!-- Sent from Reviewable.io -->
